### PR TITLE
Fix 1367

### PR DESCRIPTION
**Pull request by kamil4**
Fixes https://github.com/LycheeOrg/Lychee/issues/1367

The fix itself is a one-liner: the regular expression in Imagick mapper was not allowing a `0` for degrees/minutes/seconds. However, the way it was failing (with a type mismatch) uncovered errors in type annotations of arguments of some generic `set` methods.

### CODE REVIEW

**Review by d7415 — APPROVED** [review]
I've only *properly* looked at the regex, but the rest looks sensible.

**Comment by kamil4** [comment]
Additional testing revealed that the Imagick extractor failed to extract width/height, so I was compelled to fix that as well. Now this should really be ready.

File: lib/PHPExif/Adapter/ImageMagick.php [hunk 1]
```diff
@@ -46,7 +46,15 @@ public function getExifFromFile(string $file) : Exif
...
         $mimeType = mime_content_type($file);
-        $additional_data = array('MimeType' => $mimeType, 'filesize' => $data_filesize, 'filename' => $data_filename);
+        $data_width = $im->getImageWidth();
+        $data_height = $im->getImageHeight();
+        $additional_data = array(
+            'MimeType' => $mimeType,
+            'filesize' => $data_filesize,
+            'filename' => $data_filename,
+            'width' => $data_width,
+            'height' => $data_height);
```
Comment (nagmat84) [inline lib/PHPExif/Adapter/ImageMagick.php:56]: If we are on it, we may as well replace the legacy `array()`-function be the more efficient language construct `[]`.

```suggestion
        $additional_data = [
            'MimeType' => $mimeType,
            'filesize' => $data_filesize,
            'filename' => $data_filename,
            'width' => $data_width,
            'height' => $data_height
        ];
```

But my actual question out of curiosity is another one: Why do we need to add `width` and `height` here if it hasn't been necessary before?

File: lib/PHPExif/Exif.php [hunk 2]
```diff
@@ -969,10 +969,10 @@ public function getAltitude() : float|false
     /**
      * Sets the altitude value
      *
-     * @param string $value
+     * @param float|false $value
      * @return \PHPExif\Exif
      */
-    public function setLongitude(string $value) : Exif
+    public function setLongitude(float|false $value) : Exif
```
Comment (nagmat84) [inline lib/PHPExif/Exif.php:975]: Personally, I am not a big fan of using `false` to indicated the absence of data. I would prefer using `null` for that, i.e. `?float` in this specific case. However, I cannot estimate where a `false` value is already been used else and how much work it would be to replace it by `null` everywhere.

File: lib/PHPExif/Exif.php [hunk 3]
```diff
@@ -1023,10 +1023,10 @@ public function getLatitude() : float|false
     /**
      * Sets the imgDirection value
      *
-     * @param string $value
+     * @param float $value
      * @return \PHPExif\Exif
      */
-    public function setImgDirection(string $value) : Exif
+    public function setImgDirection(float $value) : Exif
```
Comment (nagmat84) [inline lib/PHPExif/Exif.php:1029]: Why is this a `float` only and not falsable (or even better nullable) like the other two attributes (latitude and longitude) above?

File: lib/PHPExif/Mapper/ImageMagick.php [hunk 4]
```diff
@@ -228,7 +226,7 @@ protected function extractGPSCoordinates(string $coordinates) : float|false
         if (is_numeric($coordinates) === true) {
             return ((float) $coordinates);
         } else {
-            $m = '!^([1-9][0-9]*\/[1-9][0-9]*), ([1-9][0-9]*\/[1-9][0-9]*), ([1-9][0-9]*\/[1-9][0-9]*)!';
+            $m = '!^([0-9]+\/[1-9][0-9]*), ([0-9]+\/[1-9][0-9]*), ([0-9]+\/[1-9][0-9]*)!';
```
Comment (nagmat84) [inline lib/PHPExif/Mapper/ImageMagick.php:229]: The regex looks good. But I had a look at the method `normalizeComponent` below and spotted another inconsistency.

If the regular expression does not match, we return `false` to indicate an error or absent of a valid value or error.

If the regular expression matches, but the components cannot be normalized into a `float`, e.g. in case of a string like `0/0` which would result into an division-by-zero error, we then return `0.0` as the default value. I wonder if we shouldn't return `false` in these cases, too.

**Review by nagmat84 — COMMENTED** [review]
Overall it looks good. I left some comments which are no real change requests but rather comments/questions out of curiosity.

File: lib/PHPExif/Mapper/ImageMagick.php (hunk 4 above)
Comment (d7415) [reply lib/PHPExif/Mapper/ImageMagick.php:229]: I'm inclined to agree. I was trying to find examples of problem values that led to that, but I can't. #14 seems to be where that was added, but I can't find a related issue in Lychee.

File: lib/PHPExif/Adapter/ImageMagick.php (hunk 1 above)
Comment (kamil4) [reply lib/PHPExif/Adapter/ImageMagick.php:58]: It goes back to the comment I made on Gitter: while this code is not new, it was only used for raw files in the past. It is now (after the merge of https://github.com/LycheeOrg/Lychee/pull/1309) used for JPEGs as well if exiftool is missing, which is a lot more common usecase. So I'm guessing that the bug was "always" there, it just wasn't noticed until now due to the relative obscurity of this code...

Comment (nagmat84) [reply lib/PHPExif/Adapter/ImageMagick.php:58]: Understood, but I wonder whether this is only a temporary problem and wether we actually should go the opposite path and remove these additional attributes entirely and consistently. Of course, this would be a strategic change. It would also be fine for me to make theses changes as contained in this PR now, because it is a quick fix and then remove them later again, if we decide for it.

Let me elaborate on what I mean.

These additional attributes we are considering here (i.e. file size, MIME type, fie name) are no true part of the EXIF information. I assume that ExifTool outputs them (together with a lot of other non-EXIF data like inode number, file permissions, etc.) and one originally decided to add these data for the other EXIF readers as well for consistency reasons. But actually, we are extremely inconsistent about them when it comes to spelling (i.e. capitalization, hyphenation.)

Here (in `Adapter/ImageMagick`) we use `MimeType`, `filesize` and `filename`. In  `Adapter/Native` we use camel-case everywhere as in `MimeType`, `FileSize` and `FileName`. And in `Adapter/FFProbe` we use a different variant again. That is actually the reason why am convinced that unstructured arrays to exchange data are a really bad thing.

Adding these additional attributes as part of the EXIF extraction requires extra file system calls. In order to extract the image height and with we actually load the image. The Imagick adapter loads the image via Imagick (naturally) and the native adapter uses GD image operations, i.e. `getimage`. This is something which we do anyway in Lychee to auto-rotate the image and generate the smaller image variants. As the file size, width and height may have changed after that, we re-read these attributes and do not rely on the values provided by this library.

When I added the myriad a tests as part of my PR #1351, I also added non-typical image files (TIFF, WEBP and raw examples) and also let the whole test bench run without Imagick. See here

 - [`PhotosAddHandlerTestAbstract`](https://github.com/LycheeOrg/Lychee/blob/use_filestreams/tests/Feature/PhotosAddHandlerTestAbstract.php); instantiated by
    - [`PhotosAddHandlerImagickTest`](https://github.com/LycheeOrg/Lychee/blob/use_filestreams/tests/Feature/PhotosAddHandlerImagickTest.php)
    - [`PhotosAddHandlerGDTest`](https://github.com/LycheeOrg/Lychee/blob/use_filestreams/tests/Feature/PhotosAddHandlerGDTest.php)
 - [`PhotosRotateTestAbstract`](https://github.com/LycheeOrg/Lychee/blob/use_filestreams/tests/Feature/PhotosRotateTestAbstract.php); instantiated by
    - [`PhotosRotateImagickTest`](https://github.com/LycheeOrg/Lychee/blob/use_filestreams/tests/Feature/PhotosRotateImagickTest.php)
    - [`PhotosRotateGDTest`](https://github.com/LycheeOrg/Lychee/blob/use_filestreams/tests/Feature/PhotosRotateGDTest.php)

(BTW: This PR is still pending for review, no pressure :smile:) This increased test coverage to 92% :tada: and unveiled a lot, I mean a LOT, of lurking errors which were unrelated to the actual PR. But hopefully, these tests will prevent regressions bugs like #1367 from happening again in the future, because every combination of photo type and raw image with an EXIF handler is now tested. (Embrace yourself for much longer test run-times than before). However, I fixed these problems on the Lychee side. For example, Lychee does not rely on these attributes as part of the EXIF data anymore.

In summary, I would suggest to do the following:

 - We add these additional attributes as part of this PR. This provide us with a quick fix for #1367 on the current master. I guess PR #1351 still requires some time to get merged.
 - In the long run I would rather prefer
    - to remove this additional non-truly EXIF attributes from the EXIF library, i.e. instead of adding them because EXIFtool does, I would remove them everywhere
    - replacing this unfortunate, non type-safe array `$data` by something more reliable

File: lib/PHPExif/Exif.php (hunk 2 above)
Comment (ildyria) [reply lib/PHPExif/Exif.php]: Another reason to go for `null|float` is that it allows to use `??` instead of the Elvis operator. That for me alone is a big plus.

**Review by ildyria — APPROVED** [review]
LGTM.
